tests/SpecTests/CrudSpecTest.php

```diff
@@ -0,0 +1,162 @@
+        'aggregate-out-readConcern: readConcern majority with out stage' => 'PHPLIB-431',
+        'aggregate-out-readConcern: readConcern local with out stage' => 'PHPLIB-431',
+        'aggregate-out-readConcern: readConcern available with out stage' => 'PHPLIB-431',
+        'aggregate-out-readConcern: readConcern linearizable with out stage' => 'PHPLIB-431',
+        'aggregate-out-readConcern: invalid readConcern with out stage' => 'PHPLIB-431',
+        'bulkWrite-arrayFilters: BulkWrite with arrayFilters' => 'Fails due to command assertions',
+        'updateWithPipelines: UpdateOne using pipelines' => 'PHPLIB-418',
+        'updateWithPipelines: UpdateMany using pipelines' => 'PHPLIB-418',
+        'updateWithPipelines: FindOneAndUpdate using pipelines' => 'PHPLIB-418',
+    ];
+
+    /**
+     * Assert that the expected and actual command documents match.
+     *
+     * Note: this method may modify the $expected object.
+     *
+     * @param stdClass $expected Expected command document
+     * @param stdClass $actual   Actual command document
+     */
+    public static function assertCommandMatches(stdClass $expected, stdClass $actual)
+    {
+        /* The command may contain a session id, even though we don't start one.
+         * In this case, ignore any session identifier by removing it. */
+        if (isset($actual->lsid)) {
+            unset($actual->lsid);
+        }
+
+        static::assertDocumentsMatch($expected, $actual);
+    }
+
+    /**
+     * Execute an individual test case from the specification.
+     *
+     * @dataProvider provideTests
+     * @param string    $name           Test name
+     * @param stdClass  $test           Individual "tests[]" document
+     * @param array     $runOn          Top-level "runOn" array with server requirements
+     * @param array     $data           Top-level "data" array to initialize collection
+     * @param string    $databaseName   Name of database under test
+     * @param string    $collectionName Name of collection under test
+     */
+    public function testCrud($name, stdClass $test, array $runOn = null, array $data, $databaseName = null, $collectionName = null)
+    {
+        if (isset(self::$incompleteTests[$name])) {
+            $this->markTestIncomplete(self::$incompleteTests[$name]);
+        }
+
+        if (isset($runOn)) {
+            $this->checkServerRequirements($runOn);
+        }
+
+        if (isset($test->skipReason)) {
+            $this->markTestSkipped($test->skipReason);
+        }
+
+        $databaseName = isset($databaseName) ? $databaseName : $this->getDatabaseName();
+        $collectionName = isset($collectionName) ? $collectionName : $this->getCollectionName();
+
+        $context = Context::fromCrud($test, $databaseName, $collectionName);
+        $this->setContext($context);
+
+        $this->dropTestAndOutcomeCollections();
+        $this->createTestCollection();
```

Comment:
I think this also shouldn't be needed. We only did this in the transactions spec because the server does not allow creating a collection (or database AFAIK) within a multi-document transaction.